File: infra/src/test/java/com/beat/infra/config/TaskExecutorConfigTest.java
package com.beat.infra.config;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import org.junit.jupiter.api.Test;
import org.springframework.boot.autoconfigure.AutoConfigurations;
import org.springframework.boot.autoconfigure.task.TaskExecutionAutoConfiguration;
import org.springframework.boot.test.context.runner.ApplicationContextRunner;
import org.springframework.core.task.TaskDecorator;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;

class TaskExecutorConfigTest {

	private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
		.withConfiguration(AutoConfigurations.of(TaskExecutionAutoConfiguration.class))
		.withUserConfiguration(TaskExecutorConfig.class);

	@Test
	void customizerOverridesAutoConfiguredApplicationTaskExecutorWithThreadPoolProperties() {
		contextRunner
			.withPropertyValues(
				"thread-pool.core-size=3",
				"thread-pool.max-pool-size=5",
				"thread-pool.queue-capacity=20",
				"thread-pool.thread-name-prefix=test-executor-"
			)
			.run(context -> {
				ThreadPoolProperties properties = context.getBean(ThreadPoolProperties.class);
				ThreadPoolTaskExecutor executor = context.getBean(
					"applicationTaskExecutor",
					ThreadPoolTaskExecutor.class
				);

				assertEquals(3, properties.getCoreSize());
				assertEquals(5, properties.getMaxPoolSize());
				assertEquals(20, properties.getQueueCapacity());
				assertEquals("test-executor-", properties.getThreadNamePrefix());

				assertEquals(3, executor.getCorePoolSize());
				assertEquals(5, executor.getMaxPoolSize());
				assertEquals(20, executor.getQueueCapacity());
				assertEquals("test-executor-", executor.getThreadNamePrefix());
			});
	}

	@Test
	void appliesAvailableTaskDecoratorBeansToApplicationExecutor() {
		AtomicBoolean decorated = new AtomicBoolean(false);
		TaskDecorator taskDecorator = runnable -> () -> {
			decorated.set(true);
			runnable.run();
		};

		contextRunner
			.withBean(TaskDecorator.class, () -> taskDecorator)
			.withPropertyValues(
				"thread-pool.core-size=1",
				"thread-pool.max-pool-size=1",
				"thread-pool.queue-capacity=1",
				"thread-pool.thread-name-prefix=test-executor-"
			)
			.run(context -> {
				ThreadPoolTaskExecutor executor = context.getBean(
					"applicationTaskExecutor",
					ThreadPoolTaskExecutor.class
				);
				CountDownLatch latch = new CountDownLatch(1);

				executor.execute(latch::countDown);

				assertTrue(latch.await(3, TimeUnit.SECONDS));
				assertTrue(decorated.get());
			});
	}
}

File: core/infra/src/test/kotlin/com/beat/infra/config/TaskExecutorConfigTest.kt
package com.beat.infra.config

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import org.springframework.boot.autoconfigure.AutoConfigurations
import org.springframework.boot.autoconfigure.task.TaskExecutionAutoConfiguration
import org.springframework.boot.test.context.runner.ApplicationContextRunner
import org.springframework.core.task.TaskDecorator
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class TaskExecutorConfigTest : FunSpec({
    val contextRunner =
        ApplicationContextRunner()
            .withConfiguration(AutoConfigurations.of(TaskExecutionAutoConfiguration::class.java))
            .withUserConfiguration(TaskExecutorConfig::class.java)

    test("beat async executor는 thread pool 속성으로 구성된다") {
        contextRunner
            .withPropertyValues(
                "thread-pool.core-size=3",
                "thread-pool.max-pool-size=5",
                "thread-pool.queue-capacity=20",
                "thread-pool.thread-name-prefix=test-executor-",
            ).run { context ->
                context.containsBean("applicationTaskExecutor") shouldBe true
                val properties = context.getBean(ThreadPoolProperties::class.java)
                val executor = context.getBean("beatAsyncExecutor", ThreadPoolTaskExecutor::class.java)

                properties.coreSize shouldBe 3
                properties.maxPoolSize shouldBe 5
                properties.queueCapacity shouldBe 20
                properties.threadNamePrefix shouldBe "test-executor-"
                executor.corePoolSize shouldBe 3
                executor.maxPoolSize shouldBe 5
                executor.queueCapacity shouldBe 20
                executor.threadNamePrefix shouldBe "test-executor-"
            }
    }

    test("등록된 task decorator bean이 beat async executor에 적용된다") {
        val decorated = AtomicBoolean(false)
        val taskDecorator = TaskDecorator { runnable ->
            Runnable {
                decorated.set(true)
                runnable.run()
            }
        }

        contextRunner
            .withBean(TaskDecorator::class.java, { taskDecorator })
            .withPropertyValues(
                "thread-pool.core-size=1",
                "thread-pool.max-pool-size=1",
                "thread-pool.queue-capacity=1",
                "thread-pool.thread-name-prefix=test-executor-",
            ).run { context ->
                val executor = context.getBean("beatAsyncExecutor", ThreadPoolTaskExecutor::class.java)
                val latch = CountDownLatch(1)

                executor.execute(latch::countDown)

                latch.await(3, TimeUnit.SECONDS) shouldBe true
                decorated.get() shouldBe true
            }
    }

    test("beat async executor는 기본 주입 후보에서 제외된다") {
        contextRunner
            .withPropertyValues(
                "thread-pool.core-size=3",
                "thread-pool.max-pool-size=5",
                "thread-pool.queue-capacity=20",
                "thread-pool.thread-name-prefix=test-executor-",
            ).run { context ->
                val defaultExecutor = context.getBean(Executor::class.java)
                val appExecutor = context.getBean("applicationTaskExecutor", ThreadPoolTaskExecutor::class.java)
                val beatExecutor = context.getBean("beatAsyncExecutor", ThreadPoolTaskExecutor::class.java)

                defaultExecutor shouldBe appExecutor
                (beatExecutor !== defaultExecutor) shouldBe true
            }
    }
})
